### PR TITLE
Extend CompareInsn to arrays

### LINKED ISSUE #410: Extend CompareInsn to arrays
If `T` is an ordered type, then `T[]` is an ordered type.

This is similar to #362.

Part of #328.

### PR DESCRIPTION
**Pull request by heshanpadmasiri**
## Purpose
Extend CompareInsn to arrays
Resolves #410

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: compiler/modules/front/codeGen.bal [hunk 1]
```diff
@@ -1580,20 +1584,37 @@ function operandIsNil(bir:Operand operand) returns boolean {
...
         if operandTy === t:NIL {
             return ();
         }
-        foreach bir:UniformOrderType tc in UNIFORM_ORDER_TYPES {
-            t:UniformTypeBitSet optBasicType = t:uniformTypeUnion((1 << tc) | (1 << t:UT_NIL));
-            if t:isSubtypeSimple(operandTy, optBasicType) {
-                t:UniformTypeBitSet basicType = t:uniformType(tc);
-                if t:isSubtypeSimple(operandTy, basicType) {
-                    return tc;
+        t:UniformTypeBitSet arrTy = t:uniformTypeUnion(( 1 << t:UT_LIST_RW) | (1 << t:UT_LIST_RO));
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal]: This is just t:LIST

File: compiler/modules/front/codeGen.bal [hunk 2]
```diff
@@ -1580,20 +1584,37 @@ function operandIsNil(bir:Operand operand) returns boolean {
...
+                        return <bir:ArrayOrderType> [{opt:tc}];
+                    }
+                }
+            }
+            else {
+                panic err:impossible("Failed to get array member type");
+            }
+        }
+        else {
+            foreach bir:UniformOrderType tc in UNIFORM_ORDER_TYPES {
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal]: This part should be factored out into a function and reused to get an OrderType for the member of the array.

File: compiler/modules/bir/verify.bal [hunk 3]
```diff
@@ -298,6 +300,17 @@ function verifyCompare(VerifyContext vc, CompareInsn insn) returns err:Semantic?
             operandType = t:constBasicType(operand);
         }
         if !t:isSubtypeSimple(operandType, expectType) {
+            if memberType is t:UniformTypeBitSet {
```
Comment (jclark) [inline compiler/modules/bir/verify.bal]: You should factor out the checking for the non-array case and reuse for checking the member of the array.

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/front/codeGen.bal (hunk 1 above)
Comment (heshanpadmasiri) [reply compiler/modules/front/codeGen.bal]: fixed

File: compiler/modules/front/codeGen.bal (hunk 2 above)
Comment (heshanpadmasiri) [reply compiler/modules/front/codeGen.bal]: fixed

File: compiler/modules/bir/verify.bal (hunk 3 above)
Comment (heshanpadmasiri) [reply compiler/modules/bir/verify.bal]: fixed

File: compiler/modules/nback/build.bal [hunk 4]
```diff
@@ -1106,48 +1163,67 @@ final readonly & map<bir:OrderOp> flippedOrderOps = {
...
-    llvm:BasicBlock joinBB = scaffold.addBasicBlock();
-    builder.condBr(eitherNil, eitherNilBB, neitherNilBB);
-
-    builder.positionAtEnd(eitherNilBB);
-    llvm:Value bothNil = builder.iBitwise("and", lhsIsNil, rhsIsNil);
+    bir:OrderType orderTy = insn.orderType;
+    llvm:Value? compareResult = ();
+    if orderTy is bir:OptOrderType {
+        match orderTy.opt {
+            t:UT_INT => {
```
Comment (jclark) [inline compiler/modules/nback/build.bal]: Write a function that maps UT_INT to intCompareFunction, UT_FLOAT to floatCompareFunction. Then write this statement `compareResult = builder.call(buildRuntimeFunctionDecl(scaffold, intCompareFunction), [lhs, rhs]` once using that function.

File: compiler/modules/nback/build.bal [hunk 5]
```diff
@@ -1106,48 +1163,67 @@ final readonly & map<bir:OrderOp> flippedOrderOps = {
...
-    bir:OptOrderType orderTy = <bir:OptOrderType> insn.orderType;
-    match orderTy.opt {
-        t:UT_INT => {
-            llvm:Value lhsUntagged = buildUntagInt(builder, scaffold, <llvm:PointerValue>lhs);
-            llvm:Value rhsUntagged = buildUntagInt(builder, scaffold, <llvm:PointerValue>rhs);
-            buildCompareInt(builder, scaffold, buildIntCompareOp(insn.op), lhsUntagged, rhsUntagged, result);
+    else if orderTy is bir:ArrayOrderType {
+        match orderTy[0].opt {
+            t:UT_INT => {
+                compareResult = builder.call(buildRuntimeFunctionDecl(scaffold, arrayIntCompareFunction), [lhs, rhs]);
```
Comment (jclark) [inline compiler/modules/nback/build.bal]: Similar point to above.  You can have a table with an entry for each ordered simple type e.g. {UT_INT, intCompareFunction, arrayCompareFunction}.

File: compiler/modules/nback/build.bal [hunk 6]
```diff
@@ -1106,48 +1163,67 @@ final readonly & map<bir:OrderOp> flippedOrderOps = {
...
-            llvm:Value rhsUntagged = buildUntagFloat(builder, scaffold, <llvm:PointerValue>rhs);
-            buildCompareFloat(builder, scaffold, buildFloatCompareOp(insn.op), lhsUntagged, rhsUntagged, result);
+    }
+    if compareResult is () {
+        panic error("Failed to find runtime compare function");
+    }
+    else {
+        llvm:Value? resultValue = ();
+        match insn.op {
+            ">=" => {
```
Comment (jclark) [inline compiler/modules/nback/build.bal]: Have a map that maps e.g. ">=" to "sge" (don't we have one already), and then don't repeat the statement.

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/bir/verify.bal [hunk 7]
```diff
@@ -277,8 +277,38 @@ function verifyConvertToFloatInsn(VerifyContext vc, ConvertToFloatInsn insn) ret
...
+                }
+            }
+            else {
+                panic err:impossible("Failed to get array member type");
+            }
+        }
+    }
+}
+
 function verifyCompare(VerifyContext vc, CompareInsn insn) returns err:Semantic? {
```
Comment (jclark) [inline compiler/modules/bir/verify.bal]: I think we can improve this. One minor point: please order things top down with verifyCompare first followed by the helper functions.

In verifyCompare, iterate over each operand, and compute the actual type (as you do in verifyCompareOperandTypes). For each operand, then call a function to check the that the semType of the operand is compatible with the orderType; there should separate functions F and G for list and non-list, respectively. F should call G.

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/nback/build.bal (hunk 4 above)
Comment (heshanpadmasiri) [reply compiler/modules/nback/build.bal]: fixed

File: compiler/modules/nback/build.bal (hunk 5 above)
Comment (heshanpadmasiri) [reply compiler/modules/nback/build.bal]: fixed

File: compiler/modules/nback/build.bal (hunk 6 above)
Comment (heshanpadmasiri) [reply compiler/modules/nback/build.bal]: Fixed by adding a new table to record both predicate and expected value

File: compiler/modules/bir/verify.bal (hunk 7 above)
Comment (heshanpadmasiri) [reply compiler/modules/bir/verify.bal]: Fixed

**Comment by jclark** [comment]
In the interests of time, I pushed a few tweaks myself. For future reference:

- non-const global variables need to be final and readonly
- error messages should not be capitalized

**Pull request merged**